rfc/RFC-0001-dynamic-embedding-with-PS.md

```diff
@@ -0,0 +1,95 @@
+# RFC: Dynamic Embedding with PS
+
+Authors: Zilin Zhu (zilinzhu@tencent.com), Jiayi Feng (jiayifeng@tencent.com), Yang Yu (josephyu@tencent.com)
+
+**NOTE**: This is an early draft of the proposed feature and the main focus is to make developers interested form ideal APIs and implementation for this feature (dynamic embedding). Therefore, the current doc may not conform to a certain RFC format at the moment, e.g. https://github.com/pytorch/rfcs/blob/master/RFC-0000-template.md
+
+## Summary
+In a social media scenario, it is common that the feeds to recommend and the number of users keep growing. As a result, users of TorchRec would expect their recommender system support training without estimating the upper bound of the number of feeds in advance. We would call the feature to support growing embedding size as dynamic embedding.
+
+We propose to solve dynamic embedding by attaching an external parameter server (PS) to TorchRec. TorchRec would automatically push/pull the features to/from PS during training. We also propose certain APIs for dynamic embedding, so that TorchRec could pipeline the network communication with the model computation and achieve maximum performance. 
+
+## Background
+With the rapid development of social media, recommender systems face a new challenge – the number of feeds to recommend grows daily and it is hard to train models with a fixed size embedding. Therefore, we propose a new feature to help companies train large recommendation models with dynamic embedding support and great performance.
```

Comment:
> Overall, this would still be SPMD fully synchronous training, with the exception that the PS fetches (i.e cache miss) will be async?

Yes.

> What is the mechanism responsible for allocating memory? (today, the planner produces a shardingPlan which includes which params live where, including UVM, HBM, etc). Would the planner also be responsible for allocating params across HBM (for cache) and parameter servers?

1. Would the planner also be responsible for allocating params across HBM (for cache)?
  * No. 
  * For example
      * if we want a Embedding(name="feed_ids") be infinity size/`dynamic embedding`. We just config this embedding as `Embedding(name="feed_ids", num_embeddings=65536)`. The `num_embeddings` in this case, is just the `num_embeddings` in `TorchRec Models/HBM/UVM, etc`, and we reuse TorchRec's planner to sharding it into cluster.
      * And with a decorator `DynamicEmbedding(model, remote="redis://...", embedding_names=["feed_ids"])`, we just tell DynamicEmbedding extension to swap in/out `feed_ids` Embedding.

2. allocating params in parameter servers.
  * In ParameterServer, we just implement a method like `Fetch(item_id ) -> Optional<torch::Tensor>`
  * If the fetch returns empty, then we just initialize this cache row in `TorchRec Model`.
     * something like,  `model.state_dict()["xxx.xxx.feed_ids"].local_shards().tensor[row].normal_()` or `optimizer.state_dict()[...]..`

> Would sharding still happen with torchrec.distributed.DMP?

Yes. We do not change any sharding or distributed paradigm in `TorchRec`.